pkg/compact/compact.go

```diff
@@ -0,0 +1,540 @@
+	if err != nil {
+		return errors.Wrap(err, "read meta file")
+	}
+	h := c.groupKey(meta)
+
+	c.mtx.Lock()
+	defer c.mtx.Unlock()
+
+	g, ok := c.groups[h]
+	if !ok {
+		g, err = NewGroup(
+			log.With(c.logger, "compactionGroup", h),
+			c.bkt,
+			filepath.Join(c.dir, "groups", h),
+			meta.Thanos.Labels,
+		)
+		if err != nil {
+			return errors.Wrap(err, "create compaction group")
+		}
+		c.groups[h] = g
+	}
+	return g.Add(meta)
+}
+
+// Groups returns the compaction groups created by the Syncer.
+func (c *Syncer) Groups() (res []*Group) {
+	c.mtx.Lock()
+	defer c.mtx.Unlock()
+
+	for _, g := range c.groups {
+		res = append(res, g)
+	}
+	return res
+}
+
+// GarbageCollect deletes blocks from the bucket if their data is available as part of a
+// block with a higher compaction level.
+func (c *Syncer) GarbageCollect(ctx context.Context) error {
+	// Map each block to its highest priority parent. Initial blocks have themselves
+	// in their source section, i.e. are their own parent.
+	var (
+		all     = map[ulid.ULID]*block.Meta{}
+		parents = map[ulid.ULID]ulid.ULID{}
+	)
+	err := iterMetas(c.dir, func(dir string, _ ulid.ULID) error {
+		meta, err := block.ReadMetaFile(dir)
+		if err != nil {
+			return errors.Wrap(err, "read meta")
+		}
+		all[meta.ULID] = meta
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+
+	for id, meta := range all {
+		// For each source block we contain, check whether we are the highest priotiy parent block.
+		for _, sid := range meta.Compaction.Sources {
+			pid, ok := parents[sid]
+			// No parents for the source block so far.
+			if !ok {
+				parents[sid] = id
+				continue
+			}
+			pmeta, ok := all[pid]
+			if !ok {
+				return errors.Errorf("previous parent block %s not found", pid)
+			}
+			// A parent is of higher priority if its compaction level is higher.
+			// If compaction levels are equal, the more recent ULID wins.
+			//
+			// The ULID recency alone is not sufficient since races, e.g. induced
+			// by downtime of garbage collection, may re-compact blocks that are
+			// were already compacted into higher-level blocks multiple times.
+			level, plevel := meta.Compaction.Level, pmeta.Compaction.Level
+
+			if level > plevel || (level == plevel && id.Compare(pid) > 0) {
+				parents[sid] = id
+			}
+		}
+	}
+
+	// A block can safely be deleted if they are not the highest priority parent for
+	// any source block.
+	hasChildren := map[ulid.ULID]struct{}{}
+	for _, pid := range parents {
+		hasChildren[pid] = struct{}{}
+	}
+
+	for id := range all {
+		if ctx.Err() != nil {
+			return ctx.Err()
+		}
+		if _, ok := hasChildren[id]; ok {
+			continue
+		}
+		// Spawn a new context so we always delete a block in full on shutdown.
+		delCtx, cancel := context.WithTimeout(context.Background(), 5*time.Minute)
+
+		level.Info(c.logger).Log("msg", "deleting outdated block", "block", id)
+
+		// TODO(fabxc): yet another hack due to the fuzzy directory semantics of our Buckets.
```

Comment:
just NOTE instead of TODO? This hack kind of make sense in terms of GCS, so we can append it into GCS `Delete` method as well.